packages/ponder-metadata-api/src/middleware.ts

```diff
@@ -1,127 +1,200 @@
+      const lastSyncedBlockHeight = metrics.getValue("ponder_sync_block", {
+        network,
+      });
+      let lastSyncedBlock: BlockInfo | null = null;
+      if (lastSyncedBlockHeight) {
+        lastSyncedBlock = await fetchBlockMetadata(lastSyncedBlockHeight);
+      }
+
+      // mapping last indexed block if available
       const ponderStatusForNetwork = ponderStatus.find((s) => s.network_name === network);
-      const lastIndexedBlock = mapPonderStatusBlockToBlockMetadata(ponderStatusForNetwork);
-
-      networkIndexingStatus[network] = {
-        totalBlocksCount:
-          metrics.getValue("ponder_historical_total_blocks", {
-            network,
-          }) ?? null,
-        cachedBlocksCount:
-          metrics.getValue("ponder_historical_cached_blocks", {
-            network,
-          }) ?? null,
-        lastSyncedBlock: blockInfo({
-          number:
-            metrics.getValue("ponder_sync_block", {
-              network,
-            }) ?? 0,
-          timestamp: 0,
-        }),
-        latestSafeBlock: blockInfo({
-          number: Number(latestSafeBlock.number),
-          timestamp: Number(latestSafeBlock.timestamp),
-        }),
+      let lastIndexedBlock: BlockInfo | null = null;
+      if (ponderStatusForNetwork) {
+        lastIndexedBlock = ponderBlockInfoToBlockMetadata(ponderStatusForNetwork);
+      }
+
+      networkIndexingStatusByChainId[indexedChainId] = {
+        lastSyncedBlock,
         lastIndexedBlock,
-        isRealtime: Boolean(metrics.getValue("ponder_sync_is_realtime", { network })),
-        isComplete: Boolean(metrics.getValue("ponder_sync_is_complete", { network })),
-        isQueued: lastIndexedBlock === null,
+        latestSafeBlock,
+        firstBlockToIndex: await query.firstBlockToIndexByChainId(indexedChainId, publicClient),
       } satisfies NetworkIndexingStatus;
     }
 
-    return ctx.json({
+    const response = {
       app,
-      // application dependencies version
       deps: {
-        ponder: metrics.getLabel("ponder_version_info", "version"),
-        nodejs: metrics.getLabel("nodejs_version_info", "version"),
+        ponder: formatTextMetricValue(metrics.getLabel("ponder_version_info", "version")),
+        nodejs: formatTextMetricValue(metrics.getLabel("nodejs_version_info", "version")),
       },
-      // application environment variables
       env,
-      // application runtime information
       runtime: {
-        // application build id
-        // https://github.com/ponder-sh/ponder/blob/626e524/packages/core/src/build/index.ts#L425-L431
-        codebaseBuildId: ponderMeta?.build_id,
-        // tableNames: meta?.table_names,
-        networkIndexingStatus,
+        codebaseBuildId: formatTextMetricValue(ponderMeta.build_id),
+        networkIndexingStatusByChainId,
       },
-    });
+    } satisfies MetadataMiddlewareResponse;
+
+    // validate if response is in correct state
+    validateResponse(response);
```

Comment:
This one should be useful when, for example, a breaking change lands in Ponder metrics set.